utils/frame/benchmarking-cli/src/storage/write.rs

```diff
@@ -0,0 +1,129 @@
+		H: HeaderT<Hash = H256>,
+		C: UsageProvider<Block> + HeaderBackend<Block>,
+	{
+		// Store the time that it took to write each value.
+		let mut record = BenchRecord::default();
+
+		let is_parity = matches!(cfg.database, DatabaseSource::ParityDb { path: _ });
+		let block = BlockId::Number(client.usage_info().chain.best_number);
+		let header = client.header(block)?.ok_or("Header not found")?;
+		let original_root = *header.state_root();
+		let trie = DbState::<Block>::new(storage.clone(), original_root);
+
+		info!("Preparing keys from block {}", block);
+		// Load all KV pairs and randomly shuffle them.
+		let mut kvs = trie.pairs();
+		let mut rng = Self::setup_rng();
+		kvs.shuffle(&mut rng);
+
+		info!("Writing {} keys", kvs.len());
+		// Write each value in one commit.
+		for (k, original_v) in kvs.iter() {
+			// Create a random value to overwrite with.
+			// NOTE: We use a possibly higher entropy than the original value,
+			// could be improved but acts as an over-estimation which is fine for now.
+			let mut new_v = vec![0; original_v.len()];
+			rng.fill_bytes(&mut new_v[..]);
+
+			// Interesting part here:
+			let start = Instant::now();
+			// Create a TX that will modify the Trie in the DB and
+			// calculate the root hash of the Trie after the modification.
+			let replace = vec![(k.as_ref(), Some(new_v.as_ref()))];
+			let (root, tx) = trie.storage_root(replace.iter().cloned(), self.state_version());
+			let tx = convert_tx::<Block>(tx, is_parity);
+			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
+
+			record.append(new_v.len(), start.elapsed())?;
+
+			// Now undo the change:
+			// Create a Trie with the modified root hash and replace the value with its original.
+			let trie = DbState::<Block>::new(storage.clone(), root);
+			let replace = vec![(k.as_ref(), Some(original_v.as_ref()))];
+			let (root, tx) = trie.storage_root(replace.iter().cloned(), self.state_version());
+			let tx = convert_tx::<Block>(tx, is_parity);
+			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
+
+			// Inserting the orginal value should bring us back to the original root hash.
+			if root != original_root {
+				// A crash here means hat the chain snapshot is now toast.
+				log::error!("DB corrupted. Wrong final root: {:?} vs {:?}", root, original_root);
+				std::process::exit(1);
+			}
+		}
+		Ok(record)
+	}
+}
+
+/// Converts a Trie transaction into a DB transaction.
+fn convert_tx<B: BlockT>(
+	mut tx: PrefixedMemoryDB<HashFor<B>>,
+	parity_db: bool,
+) -> Transaction<H256> {
+	let mut ret = Transaction::<H256>::default();
+
+	for (mut k, (v, rc)) in tx.drain().into_iter() {
+		// Using shorter keys is only possible for ParityDB.
+		// RocksDB needs the full key with prefix.
+		if parity_db {
+			k.drain(0..k.len() - DB_HASH_LEN);
+		}
+
+		if rc > 0 {
+			ret.set(columns::STATE, k.as_ref(), &v);
+		} else if rc < 0 {
+			ret.remove(columns::STATE, &k);
```

Comment:
I honestly don't really know if the removal should happen in the benchmarking.
In an archive node there is no removal.
In a pruned node, there will be an asynchronous pruning of the block +1000, which is ~ equivalent in term of db write, so it can make sense.

Reply:
Resetting the DB to the original state should bring more deterministic results when the benchmark is run a second time I think.  
Otherwise it would just be littered with more random data after each run, right?

Reply:
I agree with the resetting part, what I envisioned was something like:
- state A-> B only apply the 'set' , not the remove
- state B -> A (revert) apply the remove this time (sets should be useless)

Reply:
Ah ok. So you mean adding a flag to the function like `no_remove` and then first calling it with `true` and then `false`.  

But removing nodes from the trie happens when it is updated, so why do it here?  
We are only concerned about pruning nodes.

Reply:
Making it optional is even better.

> But removing nodes from the trie happens when it is updated, so why do it here?
We are only concerned about pruning nodes

The removal in the trie is writing only the insert part (changeset are stored), and the actual removal only happen when pruning (by default 1000 blocks later). In client/db/src/lib.rs there is two writes, the one for state is db_update, there is another write that is for the cache and does the removal immediatly, but its cache.

Reply:
Okay now I got it, thanks.  
So the pruning should not be part of the benchmark, will update it.  

Would it be ok for you if I do the removal part outside of the time recording?  
Since I still want to have same state before each value is written, otherwise there would some weird increase in each loop iteration.

Reply:
I don't think you need too, when you revert you can just skip the node writing and only apply the node remove.

Reply:
Wrong reasoning, on revert you need to apply both remove and insert.

Reply:
Seems to work. I split up the `set`s and `remove`s now.